### ISSUE TITLE
Resource Images and Video Sources

### ISSUE THREAD

**Issue by blainsmith** [oddnetworks/oddworks#78]
Proposing a schema change to accommodate multiple images and streams in a more flexible way.

**Resource Images**
Currently the format for resource images does not help identify information about the image for the platforms consuming them.

```js
images: {
  aspect16x9: 'http://example.com/16x9.png',
  aspect4x3: 'http://example.com/4x3.png'
}
```

The new proposed format will use an array with extra meta data per entry.

```js
images: [
  {url: 'http://example.com/16x9.png', ratio: '16:9', width: 1920, height: 1080, label: 'some-label'},
  {url: 'http://example.com/4x3.png', ratio: '4:3', width: 1024, height: 768, label: 'some-other-label'},
]
```

**Video Sources**

Currently videos only have a string set for `url` for the actual path to the video to play. In most cases this is ok, but when the web can consume the API this falls down since different browsers support different types. Since we want to support he native HTML5 `<video>` element we need to store multiple sources.

The new proposed format will allow for multiple sources and then the consuming client can chose the best option for playback.

```js
sources: [
  {url: 'http://example.com/master.m3u8', container: 'hls', mimeType: 'application/x-mpegURL'},
  {url: 'http://example.com/master.mpd', container: 'dash', mimeType: 'application/dash+xml'},
  {url: 'http://example.com/master.webm', container: 'webm', mimeType: 'video/webm'},
  {url: 'http://example.com/master.ogv', container: 'ogg', mimeType: 'video/ogg'}
]
```

This format would translate well for the `<video>` elements.

```html
<video>
  <source src="http://example.com/master.m3u8" type="application/x-mpegURL" />
  <source src="http://example.com/master.mpd" type="application/dash+xml" />
  <source src="http://example.com/master.ogv" type="video/ogg" />
  <source src="http://example.com/master.webm" type="video/webm" />
</video>
```

**Comment by lorismaz** [comment]
For my use case, I have videos with multiple rating, and multiple edited versions: Clean and TV-MA. Clean videos are videos that doesn't contains adult language, and videos that are TV-MA have 2 versions : censored and uncensored. Paying viewers can access TV-MA:uncensored version of a video. Adding a label option to the video sources could allow this.

Something like this:

```json
"rating": "TV-MA",
"sources": [
    {"url": "http://example.com/master.ogv", "container": "ogg", "mimeType": "video/ogg", "label": "censored"},
    {"url": "http://example.com/master.ogv", "container": "ogg", "mimeType": "video/ogg", "label": "uncensored"}
]
```

or something like that:

```json
"sources": [
    {"url": "http://example.com/master.ogv", "container": "ogg", "mimeType": "video/ogg", "rating": "TV-MA", "censored": false},
     {"url": "http://example.com/master.ogv", "container": "ogg", "mimeType": "video/ogg", "rating": "TV-MA", "censored": true}
]
```

The `rating` part could go in the video meta, but the censored/uncensored version of a video needs to be a different url of our video.

**Comment by pmcconnell** [comment]
I like where this is going. I might suggest following the pattern we are using for related content and adding names to each element. The SDKs can then have helper methods such as videoWithName() or even videoWithName: andContainer:  

We might also want to take it to another level so that you could have nesting

```
Images : {
   Thumbnails : {
        4_3 : { ... },
        16_9: {...}
   },
   Covers : {
         4_3 : { ... },
        16_9: {...}
   }
}
```
I think naming the content gives the most flexibility. It's more work to support but it allows most any conceivable use for these entities. 

Were you thinking of the label in this capacity? I assumed the label was for alternate display or another purpose. 

*excuse the formatting cuz iPad

**Comment by blainsmith** [comment]
@lorismaz: Very interesting case here. We haven't seen a case like this yet. This also has some user entitlement features on it so lets keep this going and get more feedback around this.

@pmcconnell: We proposed the latter to allow for arbitrary labels for users to choose instead of us defining the key in the JSON. Not sure how well the Roku could handle that, but every client can loop over an array and return an item that matches a pattern. Granted I think your version is computationally faster as O(1) since its a key lookup and mine is O(n), but with the number of items in the list it should be negligible. Were you envisioning the the user of Oddworks defining those `Thumbnail` and `Cover` key names and having our SDKs be smart enough to parse them into native language objects?

**Comment by lorismaz** [comment]
Also can be used for adding teasers to videos (not my use case, but just though about it)

```
{ url:"videourl", type:"teaser", label:"teaser-01" }
```

**Comment by pmcconnell** [comment]
Yes my scenario would require the dashboard to configure all of this. We would not provide keys. Again similar to our relationships.  

Either way can be made to work. And the computational cost differences will be negligible for the sizes of our data sets. 

I don't feel strongly about this. It was just a thought. 

This is worthy of more discussion before we implement any changes.

**Comment by brkattk** [comment]
Re: Images

I think this is the way we should go:

```
images: [
  {
    url: 'http://example.com/16x9.png',
    filename: '16x9.png',
    mimeType: 'image/png',
    ratio: '16:9',
    width: 1920,
    height: 1080,
    label: 'some-label'
  },
  // ...
]
```

`filename` and `ratio` are a bit redundant. `filename` is nice though.

I'd argue that ratio isn't necessary since you have width/height and if we're allowing arbitrary images, we wouldn't have nice strings for the ratio anyway (`16:9` mathematically is `1.777777777...`)

Minimally, you would need a `url` and `mimeType`. The rest may not be available or necessary.

**Comment by pmcconnell** [comment]
@brkattk solution is fine. the filename solves the naming issue. 

I don't think we need the ratio unless its a validated value and we can trust it.

Should we consider adding generic meta data to address some of the other requests we have seen?

**Comment by blainsmith** [comment]
I am cool with that for images too.

**Comment by blainsmith** [comment]
I think I might tackle this change in the NASA sample data so I can use this in the web channel. Anyone have any objections?

**Comment by pmcconnell** [comment]
If you need it, go for it

**Comment by blainsmith** [comment]
Closed with oddnetworks/oddworks-example-data#5
